cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java

```diff
@@ -0,0 +1,335 @@
+  protected void run() {
+    LOG.info("Start running ConstraintCheckerService");
+    if (!isRunning()) {
+      return;
+    }
+
+    for (int partition = 0; partition < getNumPartitions(); partition++) {
+      constraintCheckerThreads.add(new ConstraintCheckerThread(partition));
+    }
+
+    for (ConstraintCheckerThread thread : constraintCheckerThreads) {
+      thread.start();
+    }
+
+    for (ConstraintCheckerThread thread : constraintCheckerThreads) {
+      Uninterruptibles.joinUninterruptibly(thread);
+    }
+  }
+
+  private int getNumPartitions() {
+    try {
+      return Transactions.execute(transactional, new TxCallable<Integer>() {
+        @Override
+        public Integer call(DatasetContext context) throws Exception {
+          return getJobQueue(context).getNumPartitions();
+        }
+      });
+    } catch (TransactionFailureException e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  @Override
+  protected void triggerShutdown() {
+    LOG.info("Stopping ConstraintCheckerService.");
+    for (ConstraintCheckerThread thread : constraintCheckerThreads) {
+      thread.interrupt();
+    }
+
+    LOG.info("ConstraintCheckerService stopped.");
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    if (taskExecutorService != null) {
+      taskExecutorService.shutdownNow();
+    }
+  }
+
+  private class ConstraintCheckerThread extends Thread {
+    private final RetryStrategy scheduleStrategy;
+    private final int partition;
+    private final Deque<Job> readyJobs = new ArrayDeque<>();
+    private Job lastConsumed;
+    private int failureCount;
+
+    ConstraintCheckerThread(int partition) {
+      super(String.format("ConstraintCheckerService-%s", "id"));
+      // TODO: [CDAP-11370] Need to be configured in cdap-default.xml. Retry with delay ranging from 0.1s to 30s
+      scheduleStrategy =
+        co.cask.cdap.common.service.RetryStrategies.exponentialDelay(100, 30000, TimeUnit.MILLISECONDS);
+      this.partition = partition;
+    }
+
+
+    @Override
+    public void run() {
+      while (isRunning()) {
+        try {
+          long sleepTime = checkJobQueue();
+          // Don't sleep if sleepTime returned is 0
+          if (sleepTime > 0) {
+            TimeUnit.MILLISECONDS.sleep(sleepTime);
+          }
+        } catch (InterruptedException e) {
+          // sleep is interrupted, just exit without doing anything
+        }
+      }
+    }
+
+    /**
+     * Check jobs in job queue for constraint satisfaction.
+     *
+     * @return sleep time in milliseconds before next fetch
+     */
+    private long checkJobQueue() {
+      boolean emptyFetch = false;
+      try {
+        emptyFetch = Transactions.execute(transactional, new TxCallable<Boolean>() {
+          @Override
+          public Boolean call(DatasetContext context) throws Exception {
+            return checkJobConstraints(context);
+          }
+        });
+        failureCount = 0;
+      } catch (Exception e) {
```

Comment:
Interrupts from stopping the Service are handled via the `isRunning` call in the Thread's main `while` loop in its `run` method.